test/worker_queue_test.dart

```diff
@@ -0,0 +1,312 @@
+            return index + 1;
+          }),
+          onTaskCompleted: () async {
+            Logs().i('task${index + 1} completed');
+            await Future.delayed(const Duration(seconds: 1));
+            completedTasks.add(index + 1);
+          },
+        );
+      });
+
+      final workerQueue = DownloadWorkerQueue();
+      tasks.forEach(workerQueue.addTask);
+
+      await Future.delayed(const Duration(seconds: 9));
+      // Verify that all tasks completed in the expected order
+      expect(alreadyRunTasks, containsAllInOrder([1, 2, 3, 4]));
+      await Future.delayed(const Duration(seconds: 4));
+      expect(completedTasks, containsAllInOrder([1, 2, 3, 4]));
+    });
+
+    test("""
+      WHEN add 4 tasks at once to the queue using Future.wait
+      SHOULD the tasks are executed in order and the queue is empty after all tasks are completed
+    """, () async {
+      final completedTasks = <int>[];
+      final alreadyRunTasks = <int>[];
+
+      final tasks = List.generate(4, (index) {
+        return generateTask(
+          '1',
+          () async => await Future.delayed(const Duration(seconds: 2), () {
+            alreadyRunTasks.add(index + 1);
+            return index + 1;
+          }),
+          onTaskCompleted: () async {
+            Logs().i('task${index + 1} completed');
+            completedTasks.add(index + 1);
+          },
+        );
+      });
+
+      final workerQueue = DownloadWorkerQueue();
+      Future.wait(tasks.map((task) => workerQueue.addTask(task)));
+      await Future.delayed(const Duration(seconds: 1));
+      expect(workerQueue.queue.length, 3);
+      expect(workerQueue.queue.first.id, '1');
+
+      await Future.delayed(const Duration(seconds: 9));
+      // Verify that all tasks completed in the expected order
+      expect(alreadyRunTasks, containsAllInOrder([1, 2, 3, 4]));
+      expect(completedTasks, containsAllInOrder([1, 2, 3, 4]));
+    });
+
+    test("""
+      WHEN add 4 tasks to the worker queue
+      THEN task 1 is processing
+      THEN remove the last task from the queue
+      SHOULD the queue executes task except the removed task
+    """, () async {
+      final completedTasks = <int>[];
+      final alreadyRunTasks = <int>[];
+
+      final tasks = List.generate(4, (index) {
+        return generateTask(
+          '1',
+          () async => await Future.delayed(const Duration(seconds: 2), () {
+            alreadyRunTasks.add(index + 1);
+            return index + 1;
+          }),
+          onTaskCompleted: () async {
+            Logs().i('task${index + 1} completed');
+            completedTasks.add(index + 1);
+          },
+        );
+      });
+
+      final workerQueue = DownloadWorkerQueue();
+      Future.wait(tasks.map((task) => workerQueue.addTask(task)));
+      await Future.delayed(const Duration(seconds: 1));
+
+      workerQueue.queue.removeLast();
+      expect(workerQueue.queue.length, 2);
+      expect(workerQueue.queue.first.id, '1');
```

Comment:
still not understand what you want with it?

Reply:
update comment